src/github.com/getlantern/protected/protected.go

```diff
@@ -126,60 +130,89 @@ func Resolve(addr string) (*net.TCPAddr, error) {
 	return &net.TCPAddr{IP: ipAddr, Port: port}, nil
 }
 
-// Dial creates a new protected connection, it assumes that the address has
-// already been resolved to an IPv4 address.
+// Dial creates a new protected connection.
 // - syscall API calls are used to create and bind to the
 //   specified system device (this is primarily
 //   used for Android VpnService routing functionality)
-func Dial(network, addr string, timeout time.Duration) (net.Conn, error) {
+func (p *Protector) Dial(network, addr string, timeout time.Duration) (net.Conn, error) {
+	op := ops.Begin("protected-dial").Set("origin", addr).Set("timeout", timeout.Seconds())
+	defer op.End()
+	conn, err := p.dial(op, network, addr, timeout)
+	return conn, op.FailIf(err)
+}
+
+func (p *Protector) dial(op ops.Op, network, addr string, timeout time.Duration) (net.Conn, error) {
+	start := time.Now()
+
+	remainingTimeout := func() time.Duration {
+		return timeout - time.Now().Sub(start)
+	}
+
 	host, port, err := SplitHostPort(addr)
 	if err != nil {
 		return nil, err
 	}
 
+	_ip, timedOut, err := withtimeout.Do(remainingTimeout(), func() (interface{}, error) {
+		// do DNS query
+		ip := net.ParseIP(host)
+		if ip != nil {
+			return ip, nil
+		}
+		// Try to resolve it
+		addr, err := p.Resolve(network, addr)
```

Comment:
Since our dial function can be used in a bunch of places, I added the ability to resolve hostnames if dial was called with something other than an IP.